src/tests/ftest/telemetry/engine_events.py

```diff
@@ -160,55 +125,165 @@ def verify_events_last_events_ts(self, rank_count, restart_rank, events_last_eve
+        7. Gather final data for required metrics with the same command as in step 1.
         8. Compare the before and after values.
 
         :avocado: tags=all,full_regression
         :avocado: tags=hw,medium
         :avocado: tags=telemetry
         :avocado: tags=EngineEvents,test_engine_events
         """
-        # 1. Check that all ranks have started.
-        self.log_step("Check that all ranks have started.")
-        dmg_command = self.get_dmg_command()
-        if not check_system_query_status(data=dmg_command.system_query()):
-            self.fail("Engine failure detected at the beginning of the test!")
-
-        # 2. Gather initial data for the required metrics for all the server nodes.
+        # 1. Gather initial data for the required metrics for all the server nodes.
         self.log_step("Gather initial data for the required metrics for all the server nodes.")
         rank_count = self.server_managers[0].engines
         telemetry_before = self.collect_telemetry(rank_count=rank_count)
 
-        # 3. Stop a rank.
+        # 2. Stop a rank.
         restart_rank = rank_count - 1
         self.log_step(f"Stop rank {restart_rank} and wait for a while for RPCs to circulate")
-        dmg_command.system_stop(ranks=str(restart_rank))
-        # Wait for the RPCs to go around the ranks.
-        time.sleep(40)
+        self.server_managers[0].stop_ranks(ranks=[restart_rank], daos_log=self.d_log)
 
-        # 4. Verify the desired rank has stopped successfully.
+        # 3. Verify the desired rank has stopped successfully.
         self.log_step("Verify the desired rank has stopped successfully.")
-        self.check_rank_stopped(restart_rank=restart_rank, dmg_command=dmg_command)
+        failed_ranks = self.server_managers[0].check_rank_state(
+            ranks=[restart_rank], valid_states=["stopped", "excluded"], max_checks=15)
+        if failed_ranks:
+            self.fail(f"Rank {restart_rank} didn't stop!")
+
+        # 4. Verify that the RPCs circulated.
+        for count in range(10):
+            time.sleep(5)
+            metric_to_data = self.telemetry.get_metrics(name="engine_events_dead_ranks")
+            self.log.info("metric_to_data = %s", metric_to_data)
+
+            # Omit "engine" from the variable name for brevity. The indices correspond to ranks.
+            events_dead_ranks = [None for _ in range(rank_count)]
+            self.collect_events_dead_ranks(
+                events_dead_ranks=events_dead_ranks, metric_to_data=metric_to_data)
+
+            # Among the joined ranks, if at least one of them has 1, we conclude that it's
+            # circulated.
+            circulated = False
+            for rank in range(rank_count - 1):
+                if events_dead_ranks[rank] == 1:
+                    self.log.info("RPCs cirulated.")
+                    circulated = True
+                    break
+
+            if circulated:
+                break
```

Comment:
Just FYI `any` plus a generator here is useful

```suggestion
            if any(events_dead_ranks[rank] == 1 for rank in events_dead_ranks):
                self.log.info("RPCs cirulated.")
                break
```

Reply:
Thanks. This made the code shorter :)